src/librustc_trans/adt.rs

```diff
@@ -592,9 +573,26 @@ fn generic_type_of<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
                 }
             }
         }
+        _ => bug!("Unsupported type {} represented as {:#?}", t, l)
     }
 }
 
+fn union_fill(cx: &CrateContext, size: u64, align: u64) -> Type {
+    assert_eq!(size%align, 0);
+    let align_units = size/align;
+    match align {
+        1 => Type::array(&Type::i8(cx), align_units),
+        2 => Type::array(&Type::i16(cx), align_units),
+        4 => Type::array(&Type::i32(cx), align_units),
+        8 if machine::llalign_of_min(cx, Type::i64(cx)) == 8 =>
```

Comment:
This can use `cx.tcx().data_layout.i64_align.abi()` - in fact, look in `ty::layout` for how a similar match (for enums) was turned into a more general thing (e.g. there are architectures where `i32` doesn't have an alignment of 4).